[APM] Add ML expected model bounds as an option to Comparison controls

qn895 opened the pull request:
## Summary

This PR addresses https://github.com/elastic/kibana/issues/121520 and adds an option to show ML model bounds in the comparison controls and re-enable the expected bounds from ML anomaly detection jobs as comparison option to Latency, Throughput, and Failed Transaction Rate charts.

If user can access machine learning jobs, but there are *no** ML jobs that match the currently selected environment, the `Expected bounds` option will be shown but will be disabled:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 43 51" src="https://user-images.githubusercontent.com/43350163/169621681-e508b319-a59d-4a93-90c9-ca0ba63cda8e.png">

When user can access machine learning jobs, and there are ML jobs that match the current environment, `Expected bounds` will be enabled:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 43 40" src="https://user-images.githubusercontent.com/43350163/169621625-87f2877f-b2fc-43c7-a25a-19e427f45c8f.png">

When `Expected bounds` is selected for comparison:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 55 23" src="https://user-images.githubusercontent.com/43350163/169622282-916a6540-9267-42dc-abb7-a5337ab15cf6.png">

When switching environment, and there are no ML jobs that match the new environment, `Expected bounds` will be disabled, and the current selection will default to the first available time comparison option.

https://user-images.githubusercontent.com/43350163/169622308-b9ba323d-068f-4cc4-a897-23fdc87ccd03.mov

When a time range comparison is used, the legend for the charts will show  `Comparison: Day before` `Comparison: Week before`, or ` `Comparison: {period of time}`. Additionally, tooltip will no longer show series with N/A or null values.

<img width="1792" alt="Screen Shot 2022-05-20 at 17 56 34" src="https://user-images.githubusercontent.com/43350163/169622413-ef53c4f8-fac7-4835-8d12-d7217a141613.png">


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

qn895 committed:
[ML] Add bounds options

qn895 committed:
[ML] Add bounds options

qn895 committed:
[ML] Renable anomalyChartTimeseries boundaries

qn895 committed:
[ML] Renable anomalyChartTimeseries boundaries

qn895 committed:
[ML] Make it into string

qn895 committed:
[ML] Make it into string

qn895 committed:
[ML] Make it into string

qn895 committed:
[ML] Make it into string

qn895 committed:
Match colors

qn895 committed:
Match colors

qn895 committed:
[ML] Add comparisonEnabledRt

qn895 committed:
[ML] Add comparisonEnabledRt

qn895 committed:
[ML] Fix types, tests

qn895 committed:
[ML] Fix types, tests

qn895 committed:
[ML] Revert json bucket span change

qn895 committed:
[ML] Revert json bucket span change

qn895 committed:
Merge remote-tracking branch 'upstream/main' into apm-comparison-ml-expected-bounds

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

sqren commented:
@qn895 fyi: I rebased the PR in order to deploy the PR to a cloud instance

sqren commented on x-pack/plugins/apm/public/components/shared/charts/helper/get_chart_anomaly_timeseries.tsx:
```diff
@@ -46,7 +48,7 @@ export function getChartAnomalyTimeseries({
         },
       },
-      color: rgba(theme.eui.euiColorVis1, 0.5),
+      color: rgba(comparisonColor ?? theme.eui.euiColorVis1, 0.5),
```
Should `comparisonColor` be a boolean? And what happens if it's undefined?

sqren commented:
I noticed that it is currently possible to select "Expected bounds" when no ML jobs exist. We should either grey it out or remove the option entirely when no ML jobs exist.

dgieselaar commented on x-pack/plugins/apm/common/comparison_rt.ts:
```diff
@@ -0,0 +1,21 @@
+export type ComparisonEnabled = boolean | typeof ML_EXPECTED_BOUNDS;
+export const offsetRt = t.partial({
+  offset: t.string,
+});
```
@sqren to give some background, originally I suggested to change `comparisonEnabled: toBooleanRt` to `comparison: t.union([ t.literal('mlBounds'), t.literal('time'), t.literal('false') ])`. However that caused a lot of tests to fail, so to keep things moving for Quynh, we figured not renaming `comparisonEnabled` to `comparison` would be the best way forward for now and have somebody from the APM UI team complete the renaming later. IMHO `comparison` deciding what comparison option should be used is still the most appealing long-term goal. We only recently replaced `comparisonType` with `offset`, I'm not sure if going back to `comparisonType` is an improvement, it feels like a step backwards.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/132456/commits/39bbc8b9eb8a4889e2da1ab3465dd3ea11071f5b

sqren commented:
###  Expected bounds are overlaid on top of time comparisons
![image](https://user-images.githubusercontent.com/209966/169296146-cb5d7aeb-171e-4c40-a104-84a8da415e47.png)

### Title in legend and value in tooltip 

![image](https://user-images.githubusercontent.com/209966/169296300-fb3e730f-d4c3-41e8-ab03-847e3dc7c197.png)

- **Legend:** The legend says "Previous period" when "Expected bounds" is selected. I think we should change the legend to say "Comparison : Expected bounds". Similarly when "Week before" is selected it should say "Comparison: Week before".
- **Tooltip**: The value in the tooltip shows N/A which is incorrect. If we have nothing to show we should hide the values with N/A.
- Expected bounds always goes to 0. Would be great if we could show the actual lower bound.

qn895 committed:
Merge upstream/main into branch

qn895 committed:
Refactor to use offset with TimeRangeComparisonEnum.ExpectedBounds

qn895 committed:
Change comparisonColor to be part of anomalySeries

qn895 commented on x-pack/plugins/apm/public/components/app/error_group_overview/error_group_list/index.tsx:
```diff
@@ -61,7 +62,7 @@ interface Props {
   detailedStatisticsLoading: boolean;
   detailedStatistics: ErrorGroupDetailedStatistics;
-  comparisonEnabled?: boolean;
+  comparisonEnabled?: ComparisonEnabled;
```
Updated here https://github.com/elastic/kibana/pull/132456/commits/39bbc8b9eb8a4889e2da1ab3465dd3ea11071f5b

qn895 commented on x-pack/plugins/apm/public/components/shared/charts/failed_transaction_rate_chart/index.tsx:
```diff
@@ -171,6 +171,7 @@ export function FailedTransactionRateChart({
         yDomain={{ min: 0, max: 1 }}
         customTheme={comparisonChartTheme}
         anomalyTimeseries={preferredAnomalyTimeseries}
+        comparisonColor={previousPeriodColor}
```
Updated here https://github.com/elastic/kibana/pull/132456/commits/6091a0d4e8bd6cd1ab3fca2f84b69fc30c3f3bbb

qn895 committed:
Fix i18n

qn895 committed:
Add Comparison text to replace 'Previous period'

qn895 committed:
Fix expected bounds legend default to first

qn895 committed:
Hide values that are N/A in the tooltips

qn895 committed:
Fix i18n, color, and null values in tooltips

qn895 commented:
I have updated the PR with the following behaviors:

If user can access machine learning jobs, but there are *no** ML jobs that match the currently selected environment, the `Expected bounds` option will be shown but will be disabled:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 43 51" src="https://user-images.githubusercontent.com/43350163/169621681-e508b319-a59d-4a93-90c9-ca0ba63cda8e.png">

When user can access machine learning jobs, and there are ML jobs that match the current environment, `Expected bounds` will be enabled:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 43 40" src="https://user-images.githubusercontent.com/43350163/169621625-87f2877f-b2fc-43c7-a25a-19e427f45c8f.png">

When `Expected bounds` is selected for comparison, legend will show `Comparison: Expected bounds`:
<img width="1792" alt="Screen Shot 2022-05-20 at 17 55 23" src="https://user-images.githubusercontent.com/43350163/169622282-916a6540-9267-42dc-abb7-a5337ab15cf6.png">

When switching environment, and there are no ML jobs that match the new environment, `Expected bounds` will be disabled, and the current selection will default to the first available time comparison option.

https://user-images.githubusercontent.com/43350163/169622308-b9ba323d-068f-4cc4-a897-23fdc87ccd03.mov

When a time range comparison is used, the legend for the charts will show  `Comparison: Day before`, `Comparison: Week before`,  or `Comparison: {period of time}`. Additionally, tooltip will no longer show series with N/A or null values.

<img width="1792" alt="Screen Shot 2022-05-20 at 17 56 34" src="https://user-images.githubusercontent.com/43350163/169622413-ef53c4f8-fac7-4835-8d12-d7217a141613.png">

qn895 commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into apm-comparison-ml-expected-bounds

formgeist commented:
@qn895 Thanks for implementing this. I have gone through some of the various views and I have the following feedback.

**Comparison area styles are not consistent**

I noticed that switching between the expected bounds and time comparison, the area styles differ probably because of some configuration on the series in the chart? The fill color seems much lighter and almost unreadable for the Throughput chart. Any chance we can align the styles so they're the same?

<img width="1358" alt="CleanShot 2022-05-23 at 08 47 19@2x" src="https://user-images.githubusercontent.com/4104278/169761749-f6aa0f04-e379-4552-bc34-e687b84f65fe.png">
 
<img width="1361" alt="CleanShot 2022-05-23 at 08 47 08@2x" src="https://user-images.githubusercontent.com/4104278/169761738-467df01b-9ac0-40aa-8f7c-457f2232e0fd.png">

**Expected bounds disabled state on the overview page**

I think the user has come to expect to be able to see the available ML results directly from the Services list as we already show the health status based on the ML anomalies. I figure that we'd run into this problem by not defaulting to a single environment, so perhaps this is an even stronger case for eventually defaulting to e.g. production cc @sqren 

I wonder if we can pick in some guiding text in the disabled state (I know that's typically a no-go) but something like "Expected bounds (select environment)" could at least help the user to understand why they cannot choose this option. With the current native select box I know we have limited options.

**Services with a single environment selects the ML jobs for anomalies but not for expected bounds**

The opbeans-java service only has the `production` environment, so the ML anomaly detection feature defaults to show results from this job in the charts as shown by the anomaly plots in the chart. Could we use the same function for default to the environment for the expected bounds?

<img width="1367" alt="CleanShot 2022-05-23 at 08 49 23@2x" src="https://user-images.githubusercontent.com/4104278/169762546-521fdd8a-0de8-4eff-8304-03bbafa8ebe1.png">

<img width="1359" alt="CleanShot 2022-05-23 at 08 49 08@2x" src="https://user-images.githubusercontent.com/4104278/169762566-1dcae976-38cc-477e-8875-5df5f6515f23.png">

**Legend updates**

Thanks for adding in the legend updates. We had already previously discussed about changing the "previous period" to show the actual date / date range for the comparison to make it even easier to read. I'd just remove the "Comparison:" prefix for the options. I don't think it's needed.

Lastly, I have the same comment as @sqren - I would understand to see the expected bounds to show a lower bound instead of `0`?

peteharverson commented:
@formgeist one styling question - if I zoom in using my data set, the shaded area becomes striped. Is this styling deliberate? The stripes don't make much sense to me - the model plot data of the job is stored at `6m` intervals (the bucket span of the job) and the stripes have a frequency of `30m`.

![image](https://user-images.githubusercontent.com/7405507/169808667-9b450bb6-6e57-46fc-933c-fee04404053f.png)

formgeist commented:
> @formgeist one styling question - if I zoom in using my data set, the shaded area becomes striped. Is this styling deliberate? The stripes don't make much sense to me - the model plot data of the job is stored at `6m` intervals (the bucket span of the job) and the stripes have a frequency of `30m`.
> 
> ![image](https://user-images.githubusercontent.com/7405507/169808667-9b450bb6-6e57-46fc-933c-fee04404053f.png)

I agree that isn't necessary. I think it's from the original implementation of the bounds where missing bucket would be styled. I suggested to align the style to the existing time comparison area instead and avoid showing these.

sqren commented on x-pack/plugins/apm/public/components/shared/time_comparison/index.tsx:
```diff
@@ -31,19 +36,58 @@ export function TimeComparison() {
+  const comparisonOptions = useMemo(() => {
+    const timeComparisonOptions = getComparisonOptions({ start, end });
+
+    const hasMLJobsMatchingEnv =
+      canGetJobs &&
+      anomalyDetectionJobsStatus === 'success' &&
+      Array.isArray(anomalyDetectionJobsData?.jobs) &&
+      anomalyDetectionJobsData?.jobs.some((j) => j.environment === environment);
+
+    if (canGetJobs) {
+      timeComparisonOptions.push({
+        value: TimeRangeComparisonEnum.ExpectedBounds,
+        text: i18n.translate('xpack.apm.comparison.mlExpectedBounds', {
+          defaultMessage: 'Expected bounds',
+        }),
+        disabled: !hasMLJobsMatchingEnv,
+      });
```
1. Can we move this logic inside `getComparisonOptions`? 
2. We should not disable the option until the job has loaded. Right now if I select "Expected bounds" and then refresh the page I'm redirected back to one of the other options because "Expected bounds" is initially disabled. 
3. @dgieselaar You did a workaround so we'd show ML results even when "All" option was selected, if there was only one environment. Is the logic sharable or should it be re-implemented here?

sqren replied:
For the last question I found where we implement the logic: https://github.com/elastic/kibana/blob/a807c90310d2adf5c43694a8e4ae1f982ade0e32/x-pack/plugins/apm/common/anomaly_detection/get_preferred_service_anomaly_timeseries.ts#L30-L33

sqren replied:
Suggestion: we add the `preferedEnvironment` logic to the [environmentContext](https://github.com/elastic/kibana/blob/a807c90310d2adf5c43694a8e4ae1f982ade0e32/x-pack/plugins/apm/public/context/environments_context/environments_context.tsx).

Then usage changed like:

```diff
- const { environment, environments } = useEnvironmentsContext();
+ const { preferedEnvironment, environment, environments } = useEnvironmentsContext();
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/132456/commits/11e2b8dd0ad5dd7b2aa0b1cb8566189a259ad96b

qn895 committed:
Refactor to use preferredEnvironment

qn895 committed:
Don't disable expected bounds by default

qn895 committed:
Match color of expected bounds with time comparison color

qn895 committed:
Fix tests

qn895 committed:
Merge upstream/main into branch

qn895 committed:
Use bucket_span as minBucketSize for anomaly results

qn895 committed:
Fix previousPeriodColor undefined for latency chart

qn895 committed:
Remove 'Comparison:' in legend

qn895 committed:
Change anomalyTimeseriesColor to use currentPeriod to match stuff

qn895 commented:
Made several updates which includes:
- Refactored `preferredEnvironment` into  `useEnvironmentsContext`, use `preferredEnvironment` instead of `environment` as the indicator for what to show https://github.com/elastic/kibana/commit/11e2b8dd0ad5dd7b2aa0b1cb8566189a259ad96b
- Not disable the `Expected bounds` option by default https://github.com/elastic/kibana/commit/11e2b8dd0ad5dd7b2aa0b1cb8566189a259ad96b
- When there's no ML jobs matching the current preferred environment, the select option will show `Expected bounds (select other environment)` https://github.com/elastic/kibana/pull/132456/commits/6534e276f128a0e6adebaa681a3297d0c0e212c0
<img width="403" alt="Screen Shot 2022-05-23 at 15 29 26" src="https://user-images.githubusercontent.com/43350163/169900961-8dd76ed2-a757-4b97-b3e4-dd3dd7ea6695.png">


- Exposed `bucketSpan` from the job's analysis_config and used it as the minimum bucket size for anomaly detection time series aggregations https://github.com/elastic/kibana/pull/132456/commits/80b55d6ff0d670753e99afe8a2aa4a758860a70c
- Fixed color to be bolder/match with time comparison area charts
https://github.com/elastic/kibana/pull/132456/commits/7ccac3d35379f69920f177ab2af837504d37d1e5

<img width="1510" alt="Screen Shot 2022-05-23 at 15 23 28" src="https://user-images.githubusercontent.com/43350163/169900049-ed343295-51fa-4060-8283-eda759030c89.png">

- Removed the `Comparison: ` part in the legends  https://github.com/elastic/kibana/pull/132456/commits/6534e276f128a0e6adebaa681a3297d0c0e212c0

qn895 committed:
Fix type error

qn895 commented on x-pack/plugins/apm/public/components/routing/templates/apm_service_template/index.tsx:
```diff
@@ -127,8 +125,9 @@ function TemplateWithContext({
       <SearchBar {...searchBarOptions} />
-
-      {children}
+      <ServiceAnomalyTimeseriesContextProvider>
```
Moving ServiceAnomalyTimeseriesContextProvider to here here so that it can get updates from `useEnvironmentsContext/ApmMainTemplate`

formgeist reviewed (commented):
@qn895 Thanks for the updates to the charts and the selection of expected bounds option. I think we might want to address the select option with a customized selectable component that enables us to better support the help text. I'll open a separate ticket for this as it's not required to finish this PR.

I did notice that the Services list page does not show the expected bounds. Is that because we're not updating the Services and Dependencies list sparklines yet?

<img width="1677" alt="CleanShot 2022-05-24 at 09 23 27@2x" src="https://user-images.githubusercontent.com/4104278/169973908-f0cac3f4-81de-478e-ac20-a13966728f10.png">

<img width="1671" alt="CleanShot 2022-05-24 at 09 18 29@2x" src="https://user-images.githubusercontent.com/4104278/169973926-cfa5c80c-5ecc-4abf-91ab-2c54667df47e.png">

As noticed in the above screenshot we need to apply the configuration in the area chart that will show the connecting plots in the chart rather than the dotted line. I can't remember what it's called atm, but there should be an option in the chart config.

peteharverson commented on x-pack/plugins/apm/public/components/shared/charts/helper/get_chart_anomaly_timeseries.tsx:
```diff
@@ -36,21 +44,22 @@ export function getChartAnomalyTimeseries({
-      yAccessors: ['y0'],
-      y0Accessors: ['y1'],
+      yAccessors: ['y1'],
+      y0Accessors: ['y0'],
```
As noted by @sqren in https://github.com/elastic/kibana/pull/132456#issuecomment-1131650796, the lower bound is always being plotted as `0` so the `y0` values for `model_lower` are never getting used. Is there an extra property that needs to be set on the chart to prevent the area always extending to `0`. The `anomalyTimeseries` passed to the chart here looks ok.

peteharverson reviewed (commented):
Tested latest changes and overall looks great. Is using the lower bounds fixable in this PR?

qn895 committed:
Fix lower model bounds

qn895 committed:
Add comments

qn895 committed:
Remove fit

peteharverson reviewed (approved):
Tested and LGTM with the lower bounds of the ML model plot now displaying correctly.

I don't think we should look to display the model bounds in the sparklines shown in the Services list and Map at this stage. Those searches need to be optimized for speed for which we should investigate using sampling. Also with the limited width and height of the sparklines there may be little value in trying to show model bounds here.

sqren commented on x-pack/plugins/apm/public/components/routing/service_detail/index.tsx:
```diff
@@ -147,7 +150,7 @@ export const serviceDetail = {
               query: t.intersection([
                 t.type({
                   transactionName: t.string,
-                  comparisonEnabled: toBooleanRt,
+                  comparisonEnabled: comparisonEnabledRt,
```
I think we can just leave this unchanged. We don't need an intermediate type for `toBooleanRt`

qn895 replied:
Removed here https://github.com/elastic/kibana/pull/132456/commits/025d4eaa512d2c5f4053af629090f5fb892ed621

qn895 committed:
Remove comparisonEnabledRt

sqren commented on x-pack/plugins/apm/public/components/shared/time_comparison/get_comparison_options.ts:
```diff
@@ -120,10 +143,35 @@ export function getComparisonOptions({
+      value: TimeRangeComparisonEnum.ExpectedBounds,
+      text: disabled
+        ? i18n.translate('xpack.apm.comparison.mlExpectedBoundsDisabledText', {
+            defaultMessage: 'Expected bounds (select other environment)',
```
We should mention that anomaly detection must be enabled for the environment:

```suggestion
            defaultMessage: 'Expected bounds (anomaly detection must be enabled for env)',
```

qn895 replied:
@sqren Thanks. Updated with the new text here... I capitalized Anomaly so it looks more like this now: https://github.com/elastic/kibana/pull/132456/commits/1d889d0393896c39d68de060e837dc95ccfe9652

<img width="621" alt="Screen Shot 2022-05-24 at 16 25 59" src="https://user-images.githubusercontent.com/43350163/170136929-5352ec27-901e-4d26-af45-f7f426d885fc.png">

qn895 committed:
Change text

sorenlouv reviewed (approved):
lgtm. Thanks for all the work @qn895 . The designers and pm's might have some polish comment but I'm okay with doing that during FF.

qn895 commented:
Merging this PR as the one failed test for [the latest build](https://buildkite.com/elastic/kibana-pull-request/builds/47431) doesn't seem to be related to changes in this PR. A big thank you to the APM team for all your help 🙏

sqren commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.3|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/133825">](https://github.com/elastic/kibana/pull/133825)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.4.4 -->
